misc-utils/fadvise.c

```diff
@@ -0,0 +1,181 @@
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <fcntl.h>
+#include <getopt.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#include "c.h"
+#include "nls.h"
+#include "strutils.h"
+
+static const struct advice {
+	const char *name;
+	int num;
+} advices [] = {
+	{ "normal",     POSIX_FADV_NORMAL,     },
+	{ "sequential", POSIX_FADV_SEQUENTIAL, },
+	{ "random",     POSIX_FADV_RANDOM,     },
+	{ "noreuse",    POSIX_FADV_NOREUSE,    },
+	{ "willneeded", POSIX_FADV_WILLNEED,   },
+	{ "dontneed",   POSIX_FADV_DONTNEED,   },
+};
+
+static void __attribute__((__noreturn__)) usage(void)
+{
+	FILE *out = stdout;
+	size_t i;
+
+	fputs(USAGE_HEADER, out);
+	fprintf(out, _(" %s [options] file\n"), program_invocation_short_name);
+	fprintf(out, _(" %s [options] --fd|-d file-descriptor\n"), program_invocation_short_name);
+
+	fputs(USAGE_OPTIONS, out);
+	fputs(_(" -a, --advice <advice> applying advice to the file (default: \"dontneed\")\n"), out);
+	fputs(_(" -l, --length <num>    length for range operations, in bytes\n"), out);
+	fputs(_(" -o, --offset <num>    offset for range operations, in bytes\n"), out);
+
+	fputs(USAGE_SEPARATOR, out);
+	printf(USAGE_HELP_OPTIONS(23));
+
+	fputs(_("\nAvailable values for advice:\n"), out);
+	for (i = 0; i < ARRAY_SIZE(advices); i++) {
+		fprintf(out, "  %s\n",
+			advices[i].name);
+	}
+
+	printf(USAGE_MAN_TAIL("fadvise(1)"));
+
+	exit(EXIT_SUCCESS);
+}
+
+struct fadvise_args {
+	int fd;
+	off_t offset;
+	off_t len;
+	int advice;
+};
+
+static int fadvise_args(struct fadvise_args *args)
+{
+	return posix_fadvise(args->fd,
+			     args->offset, args->len,
+			     args->advice);
+}
+
+/* Copied from fallocate.c. */
+static loff_t cvtnum(char *s)
+{
+	uintmax_t x;
+
+	if (strtosize(s, &x))
+		return -1LL;
+
+	return x;
+}
+
+int main(int argc, char ** argv)
+{
+	int c;
+	int rc;
+	bool do_close = false;
+
+	struct fadvise_args args = {
+		.fd     = -1,
+		.offset = 0,
+		.len    = 0,
+		.advice = POSIX_FADV_DONTNEED,
+	};
+
+	static const struct option longopts[] = {
+		{ "advice",     required_argument, NULL, 'a' },
+		{ "fd",         required_argument, NULL, 'd' },
+		{ "length",     required_argument, NULL, 'l' },
+		{ "offset",     required_argument, NULL, 'o' },
+		{ "version",    no_argument,       NULL, 'V' },
+		{ "help",	no_argument,       NULL, 'h' },
+		{ NULL, 0, NULL, 0 },
+	};
+
+	setlocale(LC_ALL, "");
+	bindtextdomain(PACKAGE, LOCALEDIR);
+	textdomain(PACKAGE);
+
+	while ((c = getopt_long (argc, argv, "a:d:hl:o:", longopts, NULL)) != -1) {
+		switch (c) {
+		case 'a':
+			break;
+		case 'd':
+			args.fd = (int)cvtnum(optarg);
+			if (args.fd < 0)
+				errx(EXIT_FAILURE, _("invalid file descriptor value specified"));
+			break;
+		case 'l':
+			args.len = cvtnum(optarg);
+			if (args.len < 0)
+				errx(EXIT_FAILURE, _("invalid length value specified"));
+			break;
+		case 'o':
+			args.offset = cvtnum(optarg);
```

Comment:
What about to use  
```
   args.offset = strtosize_or_err(optarg, _("invalid offset value specified"));
```

Reply:
I replaced all `cvtnum` with `strto*_or_err` functions.